src/message_conversions.rs

```diff
@@ -0,0 +1,530 @@
+/********************************************************************************
+ * Copyright (c) 2023 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+use crate::transport::{AUTHORITY_NAME, ME_REQUEST_CORRELATION, UE_REQUEST_CORRELATION};
+use crate::{
+    create_request_id, retrieve_session_id, split_u32_to_u16, split_u32_to_u8, EventId, InstanceId,
+    ServiceId, ME_AUTHORITY,
+};
+use cxx::UniquePtr;
+use lazy_static::lazy_static;
+use log::trace;
+use protobuf::Enum;
+use std::collections::HashSet;
+use std::time::Duration;
+use tokio::sync::RwLock;
+use up_rust::{UCode, UMessage, UMessageBuilder, UMessageType, UPayloadFormat, UStatus, UUri};
+use vsomeip_sys::glue::{
+    make_message_wrapper, make_payload_wrapper, ApplicationWrapper, MessageWrapper, RuntimeWrapper,
+};
+use vsomeip_sys::safe_glue::{
+    get_data_safe, get_message_payload, get_pinned_application, get_pinned_message_base,
+    get_pinned_payload, get_pinned_runtime, offer_single_event_safe, set_data_safe,
+    set_message_payload,
+};
+use vsomeip_sys::vsomeip;
+use vsomeip_sys::vsomeip::{message_type_e, ANY_MAJOR};
+
+const UP_CLIENT_VSOMEIP_FN_TAG_CONVERT_UMSG_TO_VSOMEIP_MSG: &str = "convert_umsg_to_vsomeip_msg";
+const UP_CLIENT_VSOMEIP_FN_TAG_CONVERT_VSOMEIP_MSG_TO_UMSG: &str = "convert_vsomeip_msg_to_umsg";
+
+lazy_static! {
+    static ref OFFERED_EVENTS: RwLock<HashSet<(ServiceId, InstanceId, EventId)>> =
+        RwLock::new(HashSet::new());
+}
+
+pub async fn convert_umsg_to_vsomeip_msg(
+    umsg: &UMessage,
+    application_wrapper: &mut UniquePtr<ApplicationWrapper>,
+    runtime_wrapper: &UniquePtr<RuntimeWrapper>,
+) -> Result<UniquePtr<MessageWrapper>, UStatus> {
+    let Some(source) = umsg.attributes.source.as_ref() else {
+        return Err(UStatus::fail_with_code(
+            UCode::INVALID_ARGUMENT,
+            "Message has no source UUri",
+        ));
+    };
+
+    match umsg
+        .attributes
+        .type_
+        .enum_value_or(UMessageType::UMESSAGE_TYPE_UNSPECIFIED)
+    {
+        UMessageType::UMESSAGE_TYPE_PUBLISH => {
+            let mut vsomeip_msg =
+                make_message_wrapper(get_pinned_runtime(runtime_wrapper).create_notification(true));
+            let (_instance_id, service_id) = split_u32_to_u16(source.ue_id);
+            get_pinned_message_base(&vsomeip_msg).set_service(service_id);
+            let instance_id = 1; // TODO: Setting to 1 manually for now
```

Comment:
IIRC from reading the SOME/IP spec, I _do_ need to respect the upper four bytes of the ue_id and assign the instance_id appropriately according to them, right? Tagging @stevenhartley & @int0x27

Reply:
Yes, make a helper function to parse `ue_id` as `(instance_id, service_id)`. Make sure instance_id 0x0000 is replaced with 0x0001